File: framework/languages/java/e2e-kotlin/DiscoveryRegistryHa/Server/Consumer/src/main/java/systems/zlink/e2e/kotlin/discoveryregistryha/consumer/DelayableLocationStore.java
package systems.zlink.e2e.kotlin.discoveryregistryha.consumer;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;
import java.util.function.Supplier;
import systems.zlink.framework.locationprovider.ZLinkLocationStore;
import systems.zlink.framework.locationprovider.ZLinkStoreCancellation;
import systems.zlink.framework.locationprovider.ZLinkStoreKey;
import systems.zlink.framework.locationprovider.ZLinkStoreReadResult;
import systems.zlink.framework.locationprovider.ZLinkStoreScanRequest;
import systems.zlink.framework.locationprovider.ZLinkStoreScanResult;
import systems.zlink.framework.locationprovider.ZLinkStoreWriteRequest;
import systems.zlink.framework.locationprovider.ZLinkStoreWriteResult;

public final class DelayableLocationStore implements ZLinkLocationStore {
    private final ZLinkLocationStore inner;
    private final LocationStoreDelayState delayState;

    public DelayableLocationStore(ZLinkLocationStore inner, LocationStoreDelayState delayState) {
        this.inner = inner;
        this.delayState = delayState;
    }

    private <T> CompletionStage<T> delayed(Supplier<CompletionStage<T>> action) {
        int delay = delayState.delayMilliseconds();
        if (delay <= 0) return action.get();
        return CompletableFuture.runAsync(
                () -> { },
                CompletableFuture.delayedExecutor(delay, TimeUnit.MILLISECONDS))
            .thenCompose(ignored -> action.get());
    }

    @Override public CompletionStage<ZLinkStoreReadResult> read(ZLinkStoreKey key, ZLinkStoreCancellation cancellation) { return delayed(() -> inner.read(key, cancellation)); }
    @Override public CompletionStage<ZLinkStoreWriteResult> write(ZLinkStoreWriteRequest request, ZLinkStoreCancellation cancellation) { return delayed(() -> inner.write(request, cancellation)); }
    @Override public CompletionStage<ZLinkStoreScanResult> scan(ZLinkStoreScanRequest request, ZLinkStoreCancellation cancellation) { return delayed(() -> inner.scan(request, cancellation)); }
}

File: framework/languages/dotnet/e2e/StoreFailure/Server/Consumer/DelayableLocationStore.cs
using Zlink.Framework.LocationProvider;

namespace StoreFailure.Server.Consumer;

internal sealed class LocationStoreDelayState
{
    private int _delayMilliseconds;

    public int DelayMilliseconds => Volatile.Read(ref _delayMilliseconds);

    public void SetDelay(TimeSpan delay)
    {
        var milliseconds = (int)Math.Clamp(delay.TotalMilliseconds, 0, 5000);
        Volatile.Write(ref _delayMilliseconds, milliseconds);
    }
}

/// <summary>
/// Injects the configured delay before every public location-store operation.
/// The inner store continues to own all location semantics.
/// </summary>
internal sealed class DelayableLocationStore(
    IZLinkLocationStore inner,
    LocationStoreDelayState delayState) :
    IZLinkLocationStore
{
    private async ValueTask DelayAsync(CancellationToken cancellationToken)
    {
        var delay = delayState.DelayMilliseconds;
        if (delay > 0)
            await Task.Delay(delay, cancellationToken);
    }

    public async ValueTask<ZLinkStoreReadResult> ReadAsync(
        ZLinkStoreKey key,
        CancellationToken cancellationToken = default)
    {
        await DelayAsync(cancellationToken);
        return await inner.ReadAsync(key, cancellationToken);
    }

    public async ValueTask<ZLinkStoreWriteResult> WriteAsync(
        ZLinkStoreWriteRequest request,
        CancellationToken cancellationToken = default)
    {
        await DelayAsync(cancellationToken);
        return await inner.WriteAsync(request, cancellationToken);
    }

    public async ValueTask<ZLinkStoreScanResult> ScanAsync(
        ZLinkStoreScanRequest request,
        CancellationToken cancellationToken = default)
    {
        await DelayAsync(cancellationToken);
        return await inner.ScanAsync(request, cancellationToken);
    }
}
